pkg/yttlibrary/overlay/array.go

```diff
@@ -20,6 +20,9 @@ func (o OverlayOp) mergeArrayItem(
 
 	leftIdxs, err := ann.Indexes(leftArray)
 	if err != nil {
+		if _, ok := err.(MatchAnnotationNumMatchError); ok && ann.expects.when != nil {
```

Comment:
given that we repeat `if _, ok := err.(MatchAnnotationNumMatchError); ok && ann.expects.when != nil {` and also dig into privates of ArrayItemMatchAnnotation (eg expects), i think we should probably find a better way to distinguish if we should early exit.

may be set a field on MatchAnnotationNumMatchError about it being "conditional" (ie was used with `when`) and have a little method isMatchAnnotationNumMatchErrorConditional(err)